### PR TITLE
crush: ruleset is out of scope in CrushTester::test_with_crushtool

### PR DESCRIPTION
**Pull request by choury**
Fix: http://tracker.ceph.com/issues/18890
Signed-off-by: choury chouryzhou@tencent.com

### CODE REVIEW

File: src/crush/CrushTester.cc [hunk 1]
```diff
@@ -371,10 +371,11 @@ int CrushTester::test_with_crushtool(const char *crushtool_cmd,
     "--min-x", "1",
     "--max-x", "50",
     NULL);
+  string opt_ruleset = stringify(ruleset);
   if (ruleset >= 0) {
     crushtool.add_cmd_args(
       "--ruleset",
-      stringify(ruleset).c_str(),
+      opt_ruleset.c_str(),
```
Comment (tchaikov) [inline src/crush/CrushTester.cc]: > Temporary objects are destroyed as the last step in evaluating the full-expression (1.9) that (lexically) contains the point where they were created.

see § 12.2/3, n3337 (http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2012/n3337.pdf), and http://en.cppreference.com/w/cpp/language/lifetime

**Comment by tchaikov** [comment]
this is not a bug. please reopen it if i am wrong.

**Comment by jcsp** [comment]
As Kefu says, the existing code on master is okay, but the original fix for hammer (https://github.com/ceph/ceph/pull/13357) was correct because the hammer code was putting the char* on vector of char*, whereas on master SubProcess:add_cmd_arg is taking a copy onto a vector of std::string.

So, @choury, apologies for the confusion -- looks like this does not need fixing on master, so please could you reopen a PR for hammer.  Sorry!

**Comment by tchaikov** [comment]
sorry, @choury i closed http://tracker.ceph.com/issues/18890 thinking that it was a bug in master, reopening it!

**Pull request closed without merging**